stdlib/public/TensorFlow/OpaqueHandles.swift

```diff
@@ -16,61 +16,117 @@
 
 import CTensorFlow
 
-/// `ResourceHandle` is the type used by ops and the `#tfop()` syntax to
-/// represent TensorFlow "resource" values.  It exists only to represent edges
-/// in TensorFlow graphs, and has no host-side representation (and thus no
-/// methods).
-public final class ResourceHandle {
-  private init() {
-    fatalError("ResourceHandle is a marker type that can never be created")
-  }
-}
-
-/// `VariantHandle` is the type used by ops and the `#tfop()` syntax to
-/// represent TensorFlow "variant" values.
-/// TODO: rename this source file as this is no longer "opaque".
-public final class VariantHandle {
-  public let cTensorHandle: CTensorHandle
-
+/// A class that implements common functionality for resource and variant handles.
+public class ResourceVariantHandleBase {
+	public let cTensorHandle: CTensorHandle
+	public let cIsResource: Bool
+	
   @usableFromInline
-  init(owning cTensorHandle: CTensorHandle) {
+  init(owning cTensorHandle: CTensorHandle, cIsResource: Bool) {
     self.cTensorHandle = cTensorHandle
+		self.cIsResource = cIsResource
   }
 
   deinit {
     debugLog("De-initializing TensorHandle.")
     TFE_DeleteTensorHandle(cTensorHandle)
-    debugLog("Returning from deinit of VariantHandle.")
+    debugLog("Returning from deinit of handle.")
   }
-}
 
-extension VariantHandle : TensorSendableReceivable {
-  @inlinable
-  static func receiveFromAccelerator(
-    _ computation: _TensorComputation,
-    _ tensorID: Int
-  ) -> VariantHandle {
-    debugLog("Receiving variant tensor of id \(tensorID).")
+  static public func getTFDataTypeName(_ isResource: Bool) -> String {
+    return isResource ? "resource" : "variant"
+  }
+	
+	@inlinable
+	static func receiveTensorHandleHelper(
+		_ computation: _TensorComputation,
+    _ tensorID: Int,
+	  _ isResource: Bool) -> CTensorHandle {
+		debugLog("Receiving \(getTFDataTypeName(isResource)) tensor of id \(tensorID).")
     let status = TF_NewStatus()
     let context = _ExecutionContext.global
+		let dataType = isResource ? TF_RESOURCE : TF_VARIANT
     let cTensorHandle: CTensorHandle! = TFE_DequeueNamedTensorFromCtx(
-      context.eagerContext, Int32(tensorID), TF_VARIANT, status)
+      context.eagerContext, Int32(tensorID), dataType, status)
     checkOk(status)
     TF_DeleteStatus(status)
-    debugLog("Done receiving variant tensor of id \(tensorID).")
-    return VariantHandle(owning: cTensorHandle)    
-  }
+		return cTensorHandle
+	}
 
   @inlinable
-  func sendToAccelerator(_ computation: _TensorComputation,
+  func sendToAcceleratorHelper(_ computation: _TensorComputation,
```

Comment:
The refactoring here is not necessary and makes the code unnecessarily complicated. We have plans to unify all handles to a single TensorHandle type, so the duped code will be cleared away anyways.

Reply:
I don't have a strong preference. The factored code here is still reasonably easy to follow for me. 

I agree some further refactoring will be needed when we unify with TensorHandle type, and I imagine the refactored code here is on the path towards that eventual unification (rather than being orthogonal or incompatible).

Reply:
This is orthogonal to unification because unification is deleting all of this code. `TensorHandle`'s implementation of  `TensorSendableReceivable` will be general across different dtypes. I don't see a reason to rewrite this chunk or introduce new internal interfaces here. It's nothing but introducing unnecessary hierarchies and confusion.